sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,293 @@
+                 categorical_target=False, n_neighbors=3, copy=True):
+        self.n_features_to_select = n_features_to_select
+        self.categorical_features = categorical_features
+        self.categorical_target = categorical_target
+        self.n_neighbors = n_neighbors
+        self.copy = copy
+
+    def _get_support_mask(self):
+        return self.support_
+
+    def fit(self, X, y):
+        """Find the best features according to mutual information criterion.
+
+        Parameters
+        ----------
+        X : array_like or sparse matrix, shape (n_samples, n_features)
+            Feature matrix.
+        y : array_like, shape (n_samples,)
+            Target vector.
+
+        Returns
+        -------
+        self
+        """
+        X, y = check_X_y(X, y, accept_sparse='csr',
+                         y_numeric=not self.categorical_target)
+        n_features = X.shape[1]
+
+        if self.n_features_to_select >= 1:
+            n_features_to_select = int(self.n_features_to_select)
+        elif 0 < self.n_features_to_select < 1:
+            n_features_to_select = max(
+                1, int(self.n_features_to_select * n_features))
+        else:
+            raise ValueError("`n_features_to_select` must be positive.")
+
+        if isinstance(self.categorical_features, bool):
+            categorical_features = np.empty(n_features, dtype=bool)
+            categorical_features.fill(self.categorical_features)
+        else:
+            categorical_features = np.asarray(self.categorical_features)
+
+        # Scale all variables, regardless of being categorical or not.
+        # It's simple, clear and convenient.
+        X = scale(X, with_mean=False, copy=self.copy)
+
+        relevance = np.empty(n_features)
+        redundancy = np.empty((n_features, n_features))
+
+        if issparse(X):
+            X = X.tocsc()
+            iterate_columns = _iterate_csc_columns
+        else:
+            iterate_columns = _iterate_dense_columns
+
+        for i, xi in iterate_columns(X):
+            relevance[i] = _compute_mi(xi, y, categorical_features[i],
+                                       self.categorical_target,
+                                       self.n_neighbors)
+
+            for j, xj in iterate_columns(X, i + 1):
+                redundancy[i, j] = redundancy[j, i] = _compute_mi(
+                    xi, xj, categorical_features[i],
+                    categorical_features[j], self.n_neighbors)
+
+        support = np.zeros(n_features, dtype=bool)
+        support[np.argmax(relevance)] = True
+        for i in range(n_features_to_select - 1):
+            selected = np.nonzero(support)[0]
+            candidates = np.nonzero(~support)[0]
+            D = relevance[candidates]
+            R = np.mean(redundancy[np.ix_(candidates, selected)], axis=1)
+            support[candidates[np.argmax(D - R)]] = True
+
+        self.relevance_ = relevance
+        self.redundancy_ = redundancy
```

Comment:
As mentioned in my hidden diff comment, I suggest we do away with both these attributes, unless you believe there is a pressing need to store both.

We can just store a `score_` attribute, which depending on what the user provides, does a full ranking or just the first `n_features`.

WDYT?